src/Microsoft.AspNetCore.Hosting/Internal/HostingApplicationDiagnostics.cs

```diff
@@ -0,0 +1,269 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Primitives;
+using Microsoft.Net.Http.Headers;
+
+namespace Microsoft.AspNetCore.Hosting.Internal
+{
+    internal class HostingApplicationDiagnostics
+    {
+        private static readonly double TimestampToTicks = TimeSpan.TicksPerSecond / (double)Stopwatch.Frequency;
+
+        private const string ActivityName = "Microsoft.AspNetCore.Hosting.HttpRequestIn";
+        private const string ActivityStartKey = "Microsoft.AspNetCore.Hosting.HttpRequestIn.Start";
+
+        private const string DeprecatedDiagnosticsBeginRequestKey = "Microsoft.AspNetCore.Hosting.BeginRequest";
+        private const string DeprecatedDiagnosticsEndRequestKey = "Microsoft.AspNetCore.Hosting.EndRequest";
+        private const string DeprecatedDiagnosticsUnhandledExceptionKey = "Microsoft.AspNetCore.Hosting.UnhandledException";
+
+        private const string RequestIdHeaderName = "Request-Id";
+        private const string CorrelationContextHeaderName = "Correlation-Context";
+
+        private readonly DiagnosticListener _diagnosticListener;
+        private readonly ILogger _logger;
+
+        public HostingApplicationDiagnostics(ILogger logger, DiagnosticListener diagnosticListener)
+        {
+            _logger = logger;
+            _diagnosticListener = diagnosticListener;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void BeginRequest(HttpContext httpContext, ref HostingApplication.Context context)
+        {
+            long startTimestamp;
+
+            if (HostingEventSource.Log.IsEnabled())
+            {
+                context.EventLogEnabled = true;
+                // To keep the hot path short we defer logging in this function to non-inlines
+                RecordRequestStartEventLog(httpContext);
+            }
+
+
+            if (_diagnosticListener.IsEnabled())
+            {
+                if (_diagnosticListener.IsEnabled(ActivityName, httpContext))
+                {
+                    context.Activity = StartActivity(httpContext);
+                }
+                if (_diagnosticListener.IsEnabled(DeprecatedDiagnosticsBeginRequestKey))
+                {
+                    startTimestamp = Stopwatch.GetTimestamp();
+                    RecordBeginRequestDiagnostics(httpContext, startTimestamp);
+                }
+            }
+
+            // Scope may be relevant for a different level of logging, so we always create it
+            // see: https://github.com/aspnet/Hosting/pull/944
+            // Scope can be null if logging is not on.
+            context.Scope = HostingLoggerExtensions.RequestScope(_logger, httpContext);
```

Comment:
I believe have reached a decision on https://github.com/aspnet/Hosting/pull/944  that Scopes should not be unconditional.   Thus you should put this inside a if ((_logger.IsEnabled(LogLevel.Critical) (and all other code for ILogger should go there as well (since critical -> informational).